コメントをつけられるようにするの課題提出

yokomaru opened the pull request:
# 必須要件
## 日報
- [x] 本（books）のCRUDを作ったときの復習として日報（reports）が投稿できるようにする（reportsのCRUDを作る。入力項目はタイトルと本文だけで良い。created_atの日付を日報の投稿日とする）
- [x] 日報を更新・削除できるのはその日報を投稿した本人のみ（本のCRUDは特に制限なし）

## コメント
- [x] 本と日報の各詳細画面からコメントを投稿できるようにする。
- [x] コメントを削除できるのはそのコメントを投稿した本人のみ
- [x] 投稿したコメントは本と日報の各詳細画面から確認できる。投稿した内容に加えて、投稿したユーザーの名前（未入力ならメアド）と投稿日時も表示する

## 注意点やヒントなど
- [x] コメント入力フォームはHTML5の必須チェック機能を使うと、コメント未入力で投稿された時の考慮が不要となる
- [x] scaffoldの自動生成機能を使う場合は、JSONレスポンスのように使う予定のない機能は削除した状態でコミットする
- [x] 本や日報が削除されたらコメントはどうなるか？ユーザーが退会したらコメントはどうなるか？を検討する
- [x] 日報の投稿日やコメントの投稿日時はタイムゾーンの扱いに注意する（日本時間じゃないかも？）
- [x]  i18nの機能を使って日本語表示する（英語化は考慮不要）
- [x] 不必要に高度なテクニックは使わないように！

# 歓迎要件
- [x] コメントが1件もない場合はその旨を画面に表示する
- [x]  HTML5の必須チェックが効かない場合を考慮して、バリデーションエラー時のレスポンスも返せるようにする
- [x] コメントの編集ができるようにする

## スクショ
- [x] 本と日報の各詳細画面でコメントが投稿できている様子をスクショする
- [x] rubocopとerblintをパスさせる（要スクリーンショット）

https://bootcamp.fjord.jp/products/20249

yokomaru committed:
first commit

yokomaru committed:
add: default timezone Tokyo

- ActiveRecordのタイムゾーンを明示的に設定するためconfig.time_zoneをTokyoに設定する

yokomaru committed:
add: Report

- scafoldで必要なファイルを生成
- titleとcontentにnull: false, limitを追加
- i18n r10n追加
- paginate追加
- cssを当てるためclassを追加
- headerに日報を追加
- 日報のタイトル、内容にvaidationを追加(view側、model側)

yokomaru committed:
add: has_many reports

- ユーザモデルに日報とのhas_many関連づけを作成
- edit、create、update、destroyはcurrent_user.reportsとしてインスタンス変数を作成することで他のユーザーの日報を変更する処理をさせないようにする
- 日報に投稿者を表示する
- 編集と削除ボタンは投稿者とログインユーザーが一致しない場合は表示しない

yokomaru committed:
add: Comment

- scafoldで必要なファイルを生成
- titleとcontentにnull: false, limitを追加
- r10n追加
- cssを当てるためclassを追加
- コメントのタイトル、内容にvaidationを追加(view側、model側)
- リレーションを追加

yokomaru committed:
add: book and comment association

- 書籍の詳細からコメントのCRUDができるようにbook/commentsのネストを追加
    - route
    - view
    - contoller
- i18n対応を追加

yokomaru committed:
add: user and comment association

- 追加漏れのため、userとcommentsにhas_manyを追加

yokomaru committed:
add: display no comment

- コメントが一件もない時はコメントがありませんを表示させる

yokomaru committed:
add: report and comment association

- 日報の詳細からコメントのCRUDができるようにbook/commentsのネストを追加
    - route
    - view
    - contoller
- i18n対応を追加

yokomaru committed:
update: user common view and contoller comment

- 書籍と日報の詳細からコメントを作成できるようにする
- なるべく共通化するためにview/commentsのファイルを使用する
- 不要なviewは削除する

yokomaru committed:
update: move methods to nest controllers

- booksとreportsにネストしたcontollersにdestroyとedit、upsateなどを移す
- routeも合わせて変更

yokomaru committed:
add: use polymorphic url and path

- 共通化しやすいようにpolymorphic_pathやpolymorphic_urlを使用する

yokomaru committed:
update: move some methods to comments contoller

- 処理共通化のためComment Conrollerにedit、update、deleteを追加
- newとeditで渡したいパスが異なるため、form_withを各々のパスに記載する
- commentとuserのN+1解消

yokomaru committed:
update: localize not exist comments

- コメントが一件もない時のメッセージをlocaleファイルに追加

yokomaru committed:
add: buttons partial

- current_userだったら表示するボタンの分岐の表示を減らすためにパーシャルに切り分ける

yokomaru committed:
delete: helpers

- 使用していないヘルパーを削除する

yokomaru committed:
fix: newline at the end of the file

yokomaru committed:
fix: locale no_comments

- not_existよりno_commentsの方が直感的にわかりやすいので修正
- 遅延読み込みを利用する

yokomaru committed:
fix: to standardize quotation

- renderに渡す文字列のクォーテーションをダブルに統一する

yokomaru committed:
add: simple_format comment and report content

- 改行を表示するためにsimple_formatを追加＋hメソッドでエスケープを行う

yokomaru committed:
fix: to standardize 10000

- 10000という数字の表示を10_000に統一する

yoshitsugu commented on app/controllers/books/comments_controller.rb:
```diff
@@ -0,0 +1,23 @@
+    @book = Book.find(params[:book_id])
+    @comment = @book.comments.new(comment_params)
+    @comment.user_id = current_user.id
+    respond_to do |format|
```
HTML以外の表示分けがない場合は `respond_to` 必要ありません。

yokomaru replied:
`respond_to`はリクエストで指定されたフォーマットによって処理を分けるメソッドなので、
今回は不要でした！
該当箇所を削除対応いたします。

yoshitsugu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,71 @@
+  private
+
+  # Use callbacks to share common setup or constraints between actions.
+  def set_report
+    @report = current_user.reports.find(params[:id])
+  end
```
**ここではこのままで大丈夫です**が、before_actionで値をセットするのはアプリケーションが複雑になってくるとつらくなってきます。
scaffoldで作られるものがこうなっていたと思いますし、チームの方針などによっては問題ないのですが、一応そういう問題があるということだけ認識おいてもらえればと思います！
参考: https://speakerdeck.com/shinkufencer/how-to-using-before-action-with-happy-in-rails

yokomaru replied:
資料共有ありがとうございます！とても勉強になりました。

上記の`respond_to`もそうですが、
scafoldはそのまま使うのが正しいと思うのではなく、
あくまで足場だと言うことを意識します。

`before_action`につきましても改めて理解いたしました。次の課題で意識していきたいと思います！

#### デメリット
- 処理が「実行時にインスタンス変数がセットされていること」というような**前提**に依存しがちになる
- 実行順序の管理が難しくなる

#### 適した使用方法
- ログインしているかのチェックなど、本質的な処理に辿り着くまでの横断的かつ共通的な処理を行うのに適している

yoshitsugu commented on db/migrate/20240630135808_create_comments.rb:
```diff
@@ -0,0 +1,12 @@
+class CreateComments < ActiveRecord::Migration[7.0]
+  def change
+    create_table :comments do |t|
+      t.references :commentable, null: false, foreign_key: true, polymorphic: true
```
`Cannot add a foreign key to a polymorphic relation` でこれが実行できなかったんですが、お手元では実行できましたか？polymorphicについて外部キーはたしかにできないと思うんですが、動いているんでしょうか。

yokomaru replied:
> polymorphicについて外部キーはたしかにできない

コメントいただいた通り、外部キーを設定するとエラーになる認識です。
[当時の日報](https://bootcamp.fjord.jp/reports/97429#migration%E3%82%A8%E3%83%A9%E3%83%BC)でもエラーについて調査しておりました。

schema.rbにはcommentableにforeign_keyが設定されていないことから、
上記エラーに遭遇した際、`foreign_key: true`の表記を削除しmigrateを実行した後、
commitをし忘れたのかそのまま実装を進めてしまったのだと思われます💦

今後修正後は細かくコミットして漏れないように気をつけます🙏

また、その際に`null false`も外した状態にしてしまったのかnullを許容してしまっていたので、
合わせて`null false`を追加いたしました。

yoshitsugu reviewed (commented):
コメントしました！

yokomaru committed:
fix: create comments migrate file

- polymorphic が trueの時は foreign_key: trueを外す必要があるため削除する
- 上記反映を行ったためschema fileも更新

yokomaru committed:
delete: respond_to methods in contollers

- respond_to でフォーマットを分ける必要がないため各contollerから削除する

yokomaru committed:
update: new line comments contollers

読みづらいため改行を追加

yokomaru committed:
fix: schema.rb

- 今回の修正に関係ない箇所が変わっていたので戻す

yoshitsugu commented on app/controllers/reports/comments_controller.rb:
```diff
@@ -0,0 +1,21 @@
+
+    if @comment.save
+      redirect_to report_url(@report), notice: t('controllers.common.notice_create', name: Comment.model_name.human)
+    else
+      render 'reports/show', report_id: params[:report_id], status: :unprocessable_entity
```
Booksのほうもそうですが、こっちの動きってテストしてみましたか？
手元ではエラーになりました。
![Screenshot from 2024-07-12 08-18-32](https://github.com/user-attachments/assets/28f6e745-488b-4382-b48c-9e08604e60b8)

yokomaru replied:
MissingTemplateエラーにつきまして、
コメントのvalidationエラーのチェックをするために、
`false`になった際の分岐チェックはしていたつもりだったのですが、
[buttonを別ファイルに切り出した後](https://github.com/yokomaru/fjord-books_app-2023/pull/5/commits/16d5990d222f1021233f891161955b411c9e99d3)にそのチェックが漏れていたため、エラーを見落としておりました💦
(日報機能は今回の作成課題の一部だったのでボタン類をパーシャルに切り分けましたが、
書籍の方は既存のファイルだったため対応はしておらずエラーは起きておりませんした)

今後は修正後は各パターンの実装を検証し実行可能かを確認することを徹底いたします。

上記エラーは`render "reports/buttons`とパーシャルのパスを明示的に指定するように修正いたしました。

また、ご指摘をいただき、改めて自分で追記したcontollerのif文につきまして、
全て`false`になる分岐が正常に表示されるかをチェックし、問題ないことを確認いたしました。

yoshitsugu reviewed (commented):
一点だけコメントしました

yokomaru committed:
fix: specify buttton partial path "reports/buttons"

以下レビューにて指摘https://github.com/yokomaru/fjord-books_app-2023/pull/5#discussion_r1674813778
- 日報のコメント作成失敗時にMissingTemplateエラーになる
    - 該当箇所はreports/_buttton.html.erbを読み込む場所
    - コメントのnewはreports_conetrollerのshowで行っているが、createはreports/comments_contollerで行っているため、作成に失敗すると`reports/comments/_buttons`を読み込もうとしてエラーになる
    - render メソッドに明示的にパス`reports/buttons`を渡すことで読み込みエラーを起こさないようにする

yoshitsugu reviewed (approved):
OKです！
